common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -460,47 +473,17 @@ protected final boolean runAllTasksFrom(Queue<Runnable> taskQueue) {
      * @param taskQueue the task queue to drain.
      * @return {@code true} if at least {@link Runnable#run()} was called.
      */
-    private boolean runExistingTasksFrom(Queue<Runnable> taskQueue) {
-        return taskQueue.offer(BOOKEND_TASK) ? runExistingTasksUntilBookend(taskQueue)
-                                             : runExistingTasksUntilMaxTasks(taskQueue);
-    }
-
-    private boolean runExistingTasksUntilBookend(Queue<Runnable> taskQueue) {
+    private static boolean runExistingTasksFrom(Queue<Runnable> taskQueue) {
         Runnable task = pollTaskFrom(taskQueue);
-        // null is not expected because this method isn't called unless BOOKEND_TASK was inserted into the queue, and
-        // null elements are not permitted to be inserted into the queue.
-        if (task == BOOKEND_TASK) {
-            return false;
-        }
-        for (;;) {
-            safeExecute(task);
-            task = pollTaskFrom(taskQueue);
-            // null is not expected because this method isn't called unless BOOKEND_TASK was inserted into the queue,
-            // and null elements are not permitted to be inserted into the queue.
-            if (task == BOOKEND_TASK) {
-                return true;
-            }
-        }
-    }
-
-    private boolean runExistingTasksUntilMaxTasks(Queue<Runnable> taskQueue) {
-        Runnable task = pollTaskFrom(taskQueue);
-        // BOOKEND_TASK is not expected because this method isn't called unless BOOKEND_TASK fails to be inserted into
-        // the queue, and if was previously inserted we always drain all the elements from queue including BOOKEND_TASK.
         if (task == null) {
             return false;
         }
-        int i = 0;
-        do {
+        int remaining = taskQueue.size();
```

Comment:
Should we still take `maxPendingTasks ` into account ?

`int remaining = Math.min(maxPendingTasks, taskQueue.size());`

Reply:
The queue is created with `maxPendingTasks` as its capacity so I think that shouldn't be necessary?

Reply:
maxPendingTasks is more a hint as some queue implementations may not support this. So I guess it can't harm to do this here.